libraries/chain/resource_limits.cpp

```diff
@@ -441,4 +441,136 @@ account_resource_limit resource_limits_manager::get_account_net_limit_ex( const
+                    symbol_code token_code, symbol_code purpose_code, int16_t pct) {
+                        
+    EOS_ASSERT(1 <= pct && pct <= config::_100percent, transaction_exception, "pct must be between 0.01% and 100% (1-10000)");
+    const auto* param = _db.find<stake_param_object, by_id>(token_code.value);
+    EOS_ASSERT(param, transaction_exception, "no staking for token");
+    auto purpose_symbol = param->get_purpose_symbol(token_code, purpose_code);
+    
+    const auto& agents_idx = _db.get_mutable_index<stake_agent_index>().indices().get<stake_agent_object::by_key>();
+    const auto& grants_idx = _db.get_mutable_index<stake_grant_index>().indices().get<stake_grant_object::by_key>();
+
+    auto grantor_as_agent = get_agent(purpose_symbol, agents_idx, grantor_name);
+    
+    update_proxied_traversal(now, purpose_symbol, agents_idx, grants_idx, grantor_as_agent, param->frame_length, false);
+    
+    int64_t amount = 0;
+    auto grant_itr = grants_idx.lower_bound(grant_key(purpose_symbol, grantor_name));
+    while ((grant_itr != grants_idx.end()) &&
+           (grant_itr->purpose_id   == purpose_symbol.decimals()) &&
+           (grant_itr->token_code   == purpose_symbol.to_symbol_code()) &&
+           (grant_itr->grantor_name == grantor_name))
+    {
+        if (grant_itr->agent_name == agent_name) {
+            auto to_recall = static_cast<int64_t>((static_cast<int128_t>(grant_itr->share) * pct) / config::_100percent);
+            amount = recall_proxied_traversal(purpose_symbol, agents_idx, grants_idx, grant_itr->agent_name, to_recall, grant_itr->break_fee);
+            if (grant_itr->pct || grant_itr->share > to_recall) {
+                _db.modify(*grant_itr, [&](auto& g) { g.share -= to_recall; });
+                ++grant_itr;
+            }
+            else {
+                const auto &cur_grant = *grant_itr;
+                ++grant_itr;
+                _db.remove(cur_grant);
+            }
+        }
+        else
+            ++grant_itr;
+    }
+    
+    EOS_ASSERT(amount > 0, transaction_exception, "amount to recall must be positive");
+    _db.modify(*grantor_as_agent, [&](auto& a) {
+        a.balance += amount;
+        a.proxied -= amount;
+    });
+}
+
+int64_t resource_limits_manager::recall_proxied_traversal(symbol purpose_symbol, 
+                    const AgentsIdx& agents_idx, const GrantsIdx& grants_idx, 
+                    const account_name& agent_name, int64_t share, int16_t break_fee) {
+    
+    auto agent = get_agent(purpose_symbol, agents_idx, agent_name);
+
+    EOS_ASSERT(share >= 0, transaction_exception, "SYSTEM: share can't be negative");
+    EOS_ASSERT(share <= agent->shares_sum, transaction_exception, "SYSTEM: incorrect share val");
+    if(share == 0)
+        return 0;
+    auto share_fee = static_cast<int64_t>((static_cast<int128_t>(share) * std::min(agent->fee, break_fee)) / config::_100percent);
```

Comment:
propose move safe percent calculation to some function, we use it in many places

Reply:
u r right, thx